package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+    /// commands on the build host.
+    ///
+    /// Package always builds everything, but it can be set in release mode, and
+    /// we expect the existing tools to run from 'target/debug'. Additionally,
+    // you can't run `Package` until you have actually built `omicron-package`,
+    // which `BuildMinimal` does.
+    BuildMinimal,
+
+    /// Use all subcommands from omicron-package
+    #[structopt(flatten)]
+    Package(PackageSubCommand),
+
+    /// Create an overlay directory tree for each deployment server
+    ///
+    /// Each directory tree contains unique files for the given server that will
+    /// be populated in the svc/pkg dir.
+    ///
+    /// This is a separate subcommand so that we can reconstruct overlays
+    /// without rebuilding or repackaging.
+    Overlay,
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "thing-flinger")]
+struct Args {
+    /// The path to the deployment manifest TOML file
+    #[structopt(short, long, help = "Path to deployment manifest toml file")]
+    config: PathBuf,
+
+    #[structopt(subcommand)]
+    subcommand: SubCommand,
+}
+
+/// Errors which can be returned when executing subcommands
+#[derive(Error, Debug)]
+enum FlingError {
+    #[error("Servers not listed in configuration: {0:?}")]
+    InvalidServers(Vec<String>),
+
+    /// The parameter should be the name of the argument that could not be
+    /// properly converted to a string.
+    #[error("{0} is not valid UTF-8")]
+    BadString(String),
+
+    /// Failed to rsync omicron to build host
+    #[error("Failed to sync {src} with {dst}")]
+    FailedSync { src: String, dst: String },
+}
+
+fn validate_servers(
+    chosen: &BTreeSet<String>,
+    all: &BTreeMap<String, Server>,
+) -> Result<(), FlingError> {
+    let all = all.keys().cloned().collect();
+    let diff: Vec<String> = chosen.difference(&all).cloned().collect();
+    if !diff.is_empty() {
+        Err(FlingError::InvalidServers(diff))
+    } else {
+        Ok(())
+    }
+}
+
+// TODO: run in parallel when that option is given
+fn do_exec(
+    config: &Config,
+    cmd: String,
+    servers: Option<BTreeSet<String>>,
+) -> Result<()> {
+    if servers.is_some() {
+        validate_servers(&servers.as_ref().unwrap(), &config.servers)?;
+
+        for name in &servers.unwrap() {
+            let server = &config.servers[name];
+            ssh_exec(&server, &cmd, false)?;
+        }
+    } else {
+        for (_, server) in config.servers.iter() {
+            ssh_exec(&server, &cmd, false)?;
+        }
+    }
+    Ok(())
+}
+
+fn do_sync(config: &Config) -> Result<()> {
+    let server =
+        config.servers.get(&config.builder.server).ok_or_else(|| {
+            FlingError::InvalidServers(vec![config.builder.server.clone()])
+        })?;
+    let src = config.local_source.clone() + "/";
+    let dst = format!(
+        "{}@{}:{}",
+        server.username, server.addr, config.builder.omicron_path
+    );
+    let mut cmd = Command::new("rsync");
```

Comment:
Does this require that the command is run from the `omicron` root directory? What happens if people include a relative path for `config.local_source`? It might good to ensure that `rsync` is run from a consistent location, and/or that absolute paths are always used.

Reply:
We now check for an absolute path or bail. A validation function was added that's checked after Config is parsed.